fix(chapter): fix chapter members comparison bug

harshtandiya opened the pull request:
Fix a bug in the `FOSS Chapter` doctype where the comparison of chapter members was not done correctly. This bug caused issues when checking if a member was already a team member or not; resulting in role removal of all the members of chapter. Now, the comparison is done using the `chapter_member` attribute of each member (which maps to profile id), ensuring accurate results.

## What type of PR is this? (check all applicable)
- [x] 🐛Bug Fix

## Related Issues & Docs
closes #546

harshtandiya committed:
fix(chapter): fix chapter members comparison bug

Fix a bug in the `FOSS Chapter` doctype where the comparison of chapter members was not done correctly. This bug caused issues when checking if a member was already a team member or not; resulting in role removal of all the members of chapter. Now, the comparison is done using the `chapter_member` attribute of each member, ensuring accurate results.

rahulporuri commented on fossunited/chapters/doctype/foss_chapter/foss_chapter.py:
```diff
@@ -25,7 +25,7 @@ class FOSSChapter(WebsiteGenerator):
         chapter_members: DF.Table[FOSSChapterLeadTeamMember]
         chapter_name: DF.Data
-        chapter_type: DF.Literal["City Community", "FOSS Club", "Conference"]
+        chapter_type: DF.Literal["City Community", "FOSS Club", "Conference"]  # type: ignore
```
it's not obvious why this type hint is being ignored.

harshtandiya replied:
oof, this was not supposed to be pushed. was experimenting

rahulporuri reviewed (dismissed):
@harshtandiya please add a regression unit test in this PR. It looks like we don't have any blocker PRs at the moment so please let's take the time to fix this problem right once and for all.

RiyaMathew-11 reviewed (commented):
Functionally right. PR can be approved once after the unit-test addiion

mngshm reviewed (commented):
Looks legitimate, will wait for the tests to be added.

harshtandiya committed:
fix(chapter): handle member addition in FOSSChapter before_insert

This commit modifies the `FOSSChapter` doctype's `before_insert` method to call the `handle_member_addition` function instead of the deprecated `set_member_roles` function. This change ensures that member addition is properly handled when inserting a new chapter, avoiding any potential issues with role removal.

harshtandiya committed:
test(chapter): test role assignment on create

harshtandiya committed:
test(chapter): test role assignment on member addition

This commit adds a new test case `test_role_assignment_on_member_addition` to the `TestFOSSChapter` class in `test_foss_chapter.py`. The test verifies that when a new member is added to a chapter, they are assigned the role of 'Chapter Team Member'. This ensures that the role assignment functionality is working correctly.

harshtandiya committed:
test(chapter): test role deassignment on member removal

harshtandiya committed:
fix(foss chapter): make tests have random chapter name

RiyaMathew-11 reviewed (commented):
@harshtandiya CI tests are failing on member assignment and de-assignment tests. Please check on that.

Let's also use this chance to add Faker to [pyproject.toml](https://github.com/fossunited/fossunited/blob/develop/pyproject.toml) as discussed before.

harshtandiya committed:
chore: add Faker v27.0.0 dependency

harshtandiya committed:
test(chapter): add role creation for chapter team members and chapter leads

This commit adds code to create the "Chapter Team Member" and "Chapter Lead" roles if they don't already exist. This ensures that the necessary roles are available for assigning to members of a chapter while testing

RiyaMathew-11 commented on fossunited/chapters/doctype/foss_chapter/test_foss_chapter.py:
```diff
@@ -1,8 +1,95 @@
+        if not frappe.db.exists("Role", "Chapter Lead"):
+            frappe.get_doc({"doctype": "Role", "role_name": "Chapter Lead"}).insert(ignore_permissions=True)
+
+        lead_profile = frappe.get_doc("FOSS User Profile", {"user": "test@example.com"})
```
This test@example.com is assumed to always exist right ?

RiyaMathew-11 replied:
When I ran the tests - it failed for me first, as the users didnt exist. It might be good if we create dummy users here too? For the tests and delete them in tearDown ?

harshtandiya replied:
Test users are created by frappe if tests are enabled on your bench. They are always created, unless you have explicitly deleted them

RiyaMathew-11 commented on fossunited/chapters/doctype/foss_chapter/test_foss_chapter.py:
```diff
@@ -1,8 +1,95 @@
+        # When the chapter was created, a lead was assigned to it.
+        # Then the lead should have the role of 'Chapter Lead' and 'Chapter Team Member'
+        user = frappe.db.get_value("FOSS User Profile", chapter.chapter_members[0].chapter_member, "user")
+        has_role = frappe.db.exists("Has Role", {"role": "Chapter Team Member", "parent": user})
```
I am a bit confused about 'Has Role' mention in frappe.db.exists() call. Haven't seen the usage before. Please explain this ?

harshtandiya replied:
Has Role is a child table in Frappe which houses different roles which are linked to a parent of `User`. 

If you have a `Chapter Team Member` role. then `Has Role` will have an entry with `parent` as your user.

RiyaMathew-11 reviewed (commented):
Made a couple of comments @harshtandiya - Please resolve them and we are good to go.
